# CORITES
 Core contigs iterative expansion and scaffolding pipeline
 
## Description 
The CORITES algorithm is developed for refinment of metagenomic bins. 

**Step 0:** Metagenome binning

CORITES takes bins prepared by different binners from the same metagenome which represent the same genome.
E.g., these bins are classified identically (e.g., by GTDB) or sharing significant similarity.

**Step 1:** Identification of a core set of contigs

A core set of contigs is built from contigs shared between the selected bins. 

**Step 2:** Iterative expansion of the core set of contigs

The core set is iteratively expanded by other contigs derived from the binned assembly, using information from long-reads aligned to the assembly with minimap2. 
A new contig is included in the core set if it was bridged with any of the core set contigs by the number of long reads exceeding the threshold value specified with the bridge_strength parameter. 
Usually, the iterative expansion module converged after 3-8 iterations without dramatic inflation of the initial core set. Inflation is observed only with low bridge_strength parameter values. 

**Step 3:** Scaffolding with longstitch

The expanded core set is scaffolded with longstitch using long reads. 

**Step 4:** Scaffolding with assembly graph

Obtained final bins are additionally polished through the alignment with the assembly graph generated by SPAdes. 
A pair of contigs from the bin was combined if their order can be unambiguously determined using the topology of the assembly graph. 
If such ordering exists, the two contigs were merged, and the gap between them was filled with the highest covered path in the assembly graph.


## Usage

### Iterative expansion module



